File: engine/src/redis.ts
import { createClient, RedisClientType } from "redis";
class RedisHandler {
  private client!: RedisClientType;
  private publisher!: RedisClientType;
  private static instance: RedisHandler;

  init = async () => {
    this.client = createClient();
    this.publisher = createClient();

    await this.client.connect();
    await this.publisher.connect();
  };

  static createInstance = async () => {
    if (!RedisHandler.instance) {
      RedisHandler.instance = new RedisHandler();
      await RedisHandler.instance.init();
    }
    return RedisHandler.instance;
  };

  sendApiResponse = async (engine_response: any, engine_request_id: string) => {
    await this.publisher.publish(
      engine_request_id,
      JSON.stringify(engine_response)
    );
  };

  getMessage = async () => {
    const message = await this.client.brPop("MESSAGE", 0);
    return message;
  };

  sendToDB = async (data: any) => {
    const order = this.client.lPush("DB_UPDATE", JSON.stringify(data));
    return order;
  };

  publishOrder = (market: string, payload: any) => {
    return this.publisher.publish(`ORDER:${market}`, JSON.stringify(payload));
  };

  publishTrade = (market: string, payload: any) => {
    return this.publisher.publish(`TRADE:${market}`, JSON.stringify(payload));
  };

  publishTicker = (market: string, payload: any) => {
    return this.publisher.publish(`TICKER:${market}`, JSON.stringify(payload));
  };

  publishOrderBookWithQuantity = (market: string, payload: any) => {
    return this.publisher.publish(`BOOK:${market}`, JSON.stringify(payload));
  };

  setBookWithQuantity = (market: string, payload: any) => {
    return this.client.set(`DEPTH:${market}`, JSON.stringify(payload));
  };

  saveTickerData = async (
    market: string,
    trade: {
      market: string;
      price: number;
      quantity: number;
      trade_id: string;
    }
  ) => {
    const key = `TICKER_TRADES:${market}`;
    const now = Date.now();
    const cutoff = now - 24 * 60 * 60 * 1000; // 24 hours ago

    try {
      await this.client.zAdd(key, { score: now, value: JSON.stringify(trade) });
      await this.client.zRemRangeByScore(key, 0, cutoff);

      const result = await this.client.zRange(key, 0, -1);

      const trade_arr = result.map((item) => JSON.parse(item));

      let low = Infinity;
      let high = 0;
      let volume = 0;

      for (const t of trade_arr) {
        if (t.price < low) low = t.price;
        if (t.price > high) high = t.price;
        volume += t.quantity;
      }

      const open = trade_arr[0]?.price || 0;
      const close = trade_arr[trade_arr.length - 1]?.price || 0;

      const ticker_details = {
        market,
        ticker: {
          low: low === Infinity ? "0" : String(low),
          high: String(high),
          volume: String(volume),
          open: String(open),
          close: String(close),
          lastPrice: String(close),
        },
      };

      await this.publishTicker(market, ticker_details);
    } catch (err: any) {
      console.error(
        `[CRITICAL] Failed to execute ticker save pipeline:`,
        err.message
      );
    }
  };

  addTradeToRiskRouterStream = async (market: string, trade_data: any) => {
    console.log("Adding trade to risk router stream:", trade_data, market);
    await this.client.XADD(`trade:${market}`, "*", {
      payload: JSON.stringify(trade_data),
    });
  };

  private GROUP = "engine-group";
  private CONSUMER = "engine-1"; // unique per process/replica 1-> main engine 2-> backup starts only when 1 stops

  setupConsumerGroup = async (stream_key: string) => {
    try {
      await this.client.xGroupCreate(stream_key, this.GROUP, "$", {
        MKSTREAM: true,
      });
    } catch (error: any) {
      if (!error.message?.includes("BUSYGROUP")) {
        throw error; // anything else is a real problem — don't hide it
      }
    }
  };

  consumeOrderLoop = async (
    stream_key: string,
    onOrder: (order: any, id: string) => Promise<void>
  ) => {
    console.log("engine loop started");
    await this.setupConsumerGroup(stream_key);
    while (true) {
      const result = await this.client.xReadGroup(
        this.GROUP,
        this.CONSUMER,
        { key: stream_key, id: ">" }, // '>' = only entries never delivered to this group
        { COUNT: 10, BLOCK: 5000 }
      );
      console.log(result, "--------result");

      if (!result) continue;

      for (const stream of result) {
        for (const message of stream.messages) {
          try {
            const order = JSON.parse(message.message.payload);
            await onOrder(order, message.id);
            await this.client.xAck(stream_key, this.GROUP, message.id);
          } catch (err) {
            console.error(
              `failed processing order ${message.id}, will remain pending`,
              err
            );
          }
        }
      }
    }
  };
}
export default RedisHandler;

File: engine_rust/src/redis.rs
use std::{ error::Error, time::{ SystemTime, UNIX_EPOCH } };
use redis::{ AsyncCommands, aio::ConnectionManager, streams::StreamReadOptions };
use crate::{
    config::Config,
    types::{
        db::DbRequest,
        market::{ PublishBookWithQuantity, PublishTicker, PublishTickerData, SaveTicker },
        order::{ OrderRequest, PublishOrder },
        trade::{ PublishTrade, TradeData },
    },
};
use redis::streams::{ StreamReadReply };
use std::future::Future;

#[derive(thiserror::Error, Debug)]
pub enum RedisHandlerError {
    #[error("Redis error: {0}")] Redis(#[from] redis::RedisError),
    #[error("Serialization error: {0}")] Serialization(#[from] serde_json::Error),
    #[error("Invalid UTF-8 in stream payload: {0}")] Utf8(#[from] std::str::Utf8Error),
}

pub struct RedisHandler {
    client: ConnectionManager,
    publisher: ConnectionManager,
    consumer_group: String,
    consumer_name: String,
}

impl RedisHandler {
    pub async fn init(url: &str, config: &Config) -> Result<Self, RedisHandlerError> {
        let client = redis::Client::open(url)?.get_connection_manager().await?;
        let publisher = redis::Client::open(url)?.get_connection_manager().await?;

        let consumer_group = config.consumer_group.clone();
        let consumer_name = config.consumer_name.clone();

        Ok(Self { client, publisher, consumer_group, consumer_name })
    }

    pub async fn sent_to_db(&mut self, payload: DbRequest) -> Result<(), RedisHandlerError> {
        let serialized = serde_json::to_string(&payload)?;
        let _: () = self.client.lpush("DB_UPDATE", serialized).await?;
        Ok(())
    }

    pub async fn publish_order(
        &mut self,
        market: String,
        payload: PublishOrder
    ) -> Result<(), RedisHandlerError> {
        let serialized = serde_json::to_string(&payload)?;
        let stream_key = format!("ORDER:{}", market);
        let _: () = self.publisher.publish(stream_key, serialized).await?;

        Ok(())
    }

    pub async fn publish_trade(
        &mut self,
        market: String,
        payload: PublishTrade
    ) -> Result<(), RedisHandlerError> {
        let serialized = serde_json::to_string(&payload)?;
        let stream_key = format!("TRADE:{}", market);
        let _: () = self.publisher.publish(stream_key, serialized).await?;

        Ok(())
    }

    pub async fn publish_ticker(
        &mut self,
        market: &str,
        payload: &PublishTicker
    ) -> Result<(), RedisHandlerError> {
        let serialized = serde_json::to_string(&payload)?;
        let stream_key = format!("TICKER:{}", market);
        let _: () = self.publisher.publish(stream_key, serialized).await?;

        Ok(())
    }

    pub async fn publish_book_with_quantity(
        &mut self,
        market: String,
        payload: PublishBookWithQuantity
    ) -> Result<(), RedisHandlerError> {
        let serialized = serde_json::to_string(&payload)?;
        let stream_key = format!("BOOK:{}", market);
        let _: () = self.publisher.publish(stream_key, serialized).await?;

        Ok(())
    }

    pub async fn set_book_with_quantity(
        &mut self,
        market: &str,
        payload: &PublishBookWithQuantity,
    ) -> Result<(), RedisHandlerError> {
        let serialized = serde_json::to_string(payload)?;
        let key = format!("DEPTH:{}", market);
        let _: () = self.client.set(key, serialized).await?;
        Ok(())
    }

    pub async fn add_to_risk_router_stream(
        &mut self,
        market: String,
        payload: TradeData
    ) -> Result<(), RedisHandlerError> {
        let stream_key = format!("trade:{}", market);
        let serialized = serde_json::to_string(&payload)?;

        let _: () = self.client.xadd(stream_key, "*", &[("payload", serialized)]).await?;
        Ok(())
    }

    pub async fn save_ticker_data(
        &mut self,
        market: String,
        payload: SaveTicker
    ) -> Result<(), RedisHandlerError> {
        let stream_key = format!("TICKER_TRADES:{}", market);

        let serialized = serde_json::to_string(&payload)?;
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .expect("Time went backwards")
            .as_millis() as u64;
        let cutoff = now - 24 * 60 * 60 * 1000;

        let _: () = self.client.zadd(&stream_key, serialized, now).await?;
        let _: () = self.client.zrembyscore(&stream_key, 0, &cutoff).await?;

        let result: Vec<String> = self.client.zrange(&stream_key, 0, -1).await?;
        let trade_arr: Vec<SaveTicker> = result
            .into_iter()
            .filter_map(|item| serde_json::from_str::<SaveTicker>(&item).ok())
            .collect();

        let mut low: u64 = u64::MAX;
        let mut high = 0;
        let mut volume = 0;

        for trade in &trade_arr {
            if trade.price < low {
                low = trade.price;
            }
            if trade.price > high {
                high = trade.price;
            }
            volume += trade.quantity;
        }

        let open = trade_arr
            .first()
            .map(|t| t.price)
            .unwrap_or(0);
        let close = trade_arr
            .last()
            .map(|t| t.price)
            .unwrap_or(0);
        let low = if low == u64::MAX { 0 } else { low };

        let payload = PublishTicker {
            market: &market,
            ticker: PublishTickerData {
                low,
                high,
                volume,
                open,
                close,
                last_price: close,
            },
        };

        let _: () = self.publish_ticker(&market, &payload).await?;
        Ok(())
    }

    pub async fn setup_consumer_group(
        &mut self,
        stream_key: &str,
    ) -> Result<(), RedisHandlerError> {
        let result: Result<(), redis::RedisError> = self.client.xgroup_create_mkstream(
            stream_key,
            &self.consumer_group,
            "$"
        ).await;

        match result {
            Ok(()) => Ok(()),
            Err(err) => {
                // Tolerate BUSYGROUP when the engine restarts
                if err.code() == Some("BUSYGROUP") {
                    Ok(())
                } else {
                    Err(RedisHandlerError::from(err))
                }
            }
        }
    }

    pub async fn consume_order_loop<F, Fut>(
        &mut self,
        stream_key: &str,
        mut on_order: F
    )
        -> Result<(), RedisHandlerError>
        where
            F: FnMut(OrderRequest, String) -> Fut,
            Fut: Future<Output = Result<(), Box<dyn Error>>>
    {
        println!("engine loop started");
        self.setup_consumer_group(stream_key).await?;

        loop {
            let opts = StreamReadOptions::default()
                .group(&self.consumer_group, &self.consumer_name)
                .count(10)
                .block(5000);

            let reply: Option<StreamReadReply> = self.client.xread_options(
                &[&stream_key],
                &[">"],
                &opts
            ).await?;

            let Some(reply) = reply else {
                continue;
            };

            for stream in reply.keys {
                for message in stream.ids {
                    let message_id = message.id;

                    let payload_str = match message.map.get("payload") {
                        Some(redis::Value::BulkString(bytes)) => {
                            match std::str::from_utf8(bytes) {
                                Ok(s) => s,
                                Err(err) => {
                                    eprintln!("Invalid UTF-8 in payload for {message_id}: {err}");
                                    continue;
                                }
                            }
                        }
                        _ => {
                            eprintln!("Missing 'payload' field in message {message_id}");
                            continue;
                        }
                    };

                    // Deserialize the JSON payload
                    let order: OrderRequest = match serde_json::from_str(payload_str) {
                        Ok(order) => order,
                        Err(err) => {
                            eprintln!("Failed parsing order JSON for {message_id}: {err}");
                            continue;
                        }
                    };
                    match on_order(order, message_id.clone()).await {
                        Ok(()) => {
                            let ack_res: Result<(), redis::RedisError> = self.client.xack(
                                stream_key,
                                &self.consumer_group,
                                &[&message_id]
                            ).await;

                            if let Err(ack_err) = ack_res {
                                eprintln!("Failed to XACK message {message_id}: {ack_err}");
                            }
                        }
                        Err(err) => {
                            eprintln!(
                                "Failed processing order {message_id}, will remain pending: {err}"
                            );
                        }
                    }
                }
            }
        }
    }
}
